src/puppetlabs/puppetdb/cli/shovel.clj

```diff
@@ -0,0 +1,374 @@
+            [schema.core :as s]))
+
+(def cli-description "Transfer messages from ActiveMQ to Stockpile")
+
+(defn mq-dir [config]
+  (str (io/file (get-in config [:global :vardir]) "mq")))
+
+(defn ^Session create-session [^Connection connection]
+  (.createSession connection false Session/CLIENT_ACKNOWLEDGE))
+
+(defn ^Queue create-queue [^Session session endpoint]
+  (.createQueue session endpoint))
+
+(defn ^Connection create-connection [^ConnectionFactory factory]
+  (doto (.createConnection factory)
+    (.setExceptionListener
+     (reify ExceptionListener
+       (onException [this ex]
+         (log/error ex (i18n/trs "receiver queue connection error")))))
+    (.start)))
+
+(defn extract-headers
+  "Creates a map of custom headers included in `message`, currently only
+  supports String headers."
+  [^Message msg]
+  (into {}
+        (for [k (enumeration-seq (.getPropertyNames msg))]
+          [(keyword k) (.getStringProperty msg k)])))
+
+(defprotocol ConvertMessageBody
+  (convert-message-body ^String [msg]))
+
+(extend-protocol ConvertMessageBody
+  TextMessage
+  (convert-message-body [msg]
+    (-> msg
+        .getText
+        (.getBytes "UTF-8")))
+  BytesMessage
+  (convert-message-body [msg]
+    (let [len (.getBodyLength msg)
+          buf (byte-array len)
+          n (.readBytes msg buf)]
+      (when (not= len n)
+        (throw (Exception. (i18n/trs "Only read {0}/{1} bytes from incoming message" n len))))
+      buf)))
+
+(defn coerce-clj->json-byte-stream
+  "Converts clojure data to JSON serialized bytes on a
+  ByteArrayInputStream. Cheshire will only output to writers and
+  stockpile will only accept input streams, so some conversion needs
+  to be done in this fn"
+  [payload]
+  (with-open [baos (java.io.ByteArrayOutputStream.)]
+    (with-open [osw (java.io.OutputStreamWriter. baos java.nio.charset.StandardCharsets/UTF_8)]
+      (json/generate-stream payload osw))
+    (-> baos
+        .toByteArray
+        java.io.ByteArrayInputStream.)))
+
+(defn convert-old-command-format
+  "Converts from the command format that didn't include the certname
+  in the toplevel key list and had all the command data with the
+  payload key"
+  [headers message-bytes]
+  (let [{:keys [command version] :as old-command} (json/parse-strict message-bytes true)]
+    {:command command
+     :version (Long/parseLong version)
+     :certname (or (get-in old-command [:payload :certname])
+                   (get-in old-command [:payload :name]))
+     :payload (coerce-clj->json-byte-stream (:payload old-command))}))
+
+(def upgrade-message-schema
+  {:command s/Str
+   :version s/Int
+   :certname s/Str
+   :payload java.io.ByteArrayInputStream})
+
+(s/defn coerce-to-new-command :-  upgrade-message-schema
+  "Commands coming out of this function are suitable for enqueuing
+  into stockpile. They must include command, version, certname and a
+  payload"
+  [{:keys [command version certname] :as headers} message-bytes]
```

Comment:
I don't think I agree. `convert-old-command-format` doesn't use it, but it's used twice in the function. Are you suggesting I pull those values out twice by keyword instead of once by destructuring?